ghost/core/test/integration/services/mega.test.js

```diff
@@ -94,4 +101,117 @@ describe('MEGA', function () {
             emailModel.get('status').should.eql('failed');
         });
     });
+
+    /**
+     * This is one full E2E test that tests if the tracked links are valid and are working.
+     * More detailed tests don't have to cover the whole flow. But this test is useful because it also tests if the member uuids are correctly added in every link
+     * + it tests if all the pieces glue together correctly
+     */
+    describe('Link click tracking', function () {
+        before(async function () {
+            const agents = await agentProvider.getAgentsWithFrontend();
+            agent = agents.adminAgent;
+            frontendAgent = agents.frontendAgent;
+
+            await fixtureManager.init('newsletters', 'members:newsletters');
+            await agent.loginAsOwner();
+            _sendEmailJob = require('../../../core/server/services/mega/mega')._sendEmailJob;
+            _mailgunClient = require('../../../core/server/services/bulk-email')._mailgunClient;
+        });
+
+        afterEach(function () {
+            mockManager.restore();
+        });
+
+        it('Tracks all the links in an email', async function () {
+            sinon.stub(_mailgunClient, 'getInstance').returns({});
+            const sendStub = sinon.stub(_mailgunClient, 'send');
+
+            sendStub.callsFake(async () => {
+                return {
+                    id: 'stubbed-email-id'
+                };
+            });
+
+            // Prepare a post and email model
+            const emailModel = await createPublishedPostEmail();
+
+            // Launch email job
+            await _sendEmailJob({emailModel, options: {}});
+
+            await emailModel.refresh();
+            emailModel.get('status').should.eql('submitted');
+
+            // Get email data (first argument to sendStub call)
+            const emailData = sendStub.args[0][0];
+            const recipientData = sendStub.args[0][1];
+            const replacements = sendStub.args[0][2];
+
+            const recipient = recipientData[Object.keys(recipientData)[0]];
+
+            // Do the actual replacements for the first member, so we don't have to worry about them anymore
```

Comment:
Comment confused me - this does it for all members, right?

Reply:
No, it takes the first recipient (= member) of the email above, and replaces all the replacements with the replacements defined for that member

Reply:
- `replacements` contains the ids and syntax or all the replacements (this is acutally never send to mailgun, but we stub the mailgun client which is our package)
- `recipientData ` contains all the recipients and the replacement data for that recipient

Reply:
Ahhh okay - why don't we do it "properly" and use the `index` param of `forEach` to look up the recipient?\

Not saying we should - just asking why we don't ☺️

Reply:
I'm confused 🤔

Reply:
Can you give an example of what you mean?

Reply:
Something like this:

```js
            replacements.forEach((replacement, index) => {
                 const recipient = recipientData[Object.keys(recipientData)[index]];
                 emailData[replacement.format] = emailData[replacement.format].replace(
                     replacement.match,
                     recipient[replacement.id]
                 );

                 // Also force Mailgun format
                 emailData[replacement.format] = emailData[replacement.format].replace(
                     `%recipient.${replacement.id}%`,
                     recipient[replacement.id]
                 );
             });
```

Or am I completely misunderstanding the situation here?

Reply:
The replacements index doesn't have anything to do with the recipients indexes 🤔 Also, it doesn't really make sense to replace a HTML string (intended as a template for multiple recipients, where the template will get replaced depending on who is receiving the email) with data from multiple recipients 🙃 
Note that `emailData[replacement.format]` is referring to the same HTML string.

Its like replacing the first {{firstname}} in the email with the name of the first recipient, and replacing the {{lastname}} in the same email with the lastname of the following recipient. Does that make it a bit more clear? 😊

Reply:
Okay so I've totally misunderstood 😁